packages/react-components/src/components/VideoGallery/useVideoStreamLifecycleMaintainer.ts

```diff
@@ -0,0 +1,101 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import React, { useEffect, useRef } from 'react';
+import { VideoStreamOptions, CreateVideoStreamViewResult, ScalingMode } from '../../types';
+
+/**
+ * Helper hook to maintain the video stream lifecycle. This calls onCreateStreamView and onDisposeStreamView
+ * appropriately based on react lifecycle events and prop changes.
+ * This also handles calls to view.update* appropriately such as view.updateScalingMode().
+ *
+ * @private
+ */
+export const useVideoStreamLifecycleMaintainer = (props: {
+  isStreamAvailable?: boolean;
+  renderElementExists?: boolean;
+  isMirrored?: boolean;
+  scalingMode?: ScalingMode;
+  onCreateStreamView?: (options?: VideoStreamOptions) => Promise<void | CreateVideoStreamViewResult>;
+  onDisposeStreamView?: () => void;
+}): void => {
+  const { onCreateStreamView, onDisposeStreamView, isStreamAvailable, renderElementExists, isMirrored, scalingMode } =
+    props;
+
+  // HANDLING CHANGES TO VIDEO VIEW OPTIONS
+  //
+  // When VideoViewOptions change two things may happen:
+  //
+  // 1. Just the scaling mode has changed and we have access to the original StreamRenderView
+  //   - In this case we just need to call updateScalingMode on the view and do not need to recreate the stream view
+  // 2. isMirrored has changed, or scalingMode change and we don't have access to the original StreamRenderView
+  //   - In this case we need to dispose the old view and create a new one
+  //
+  // For scenario 1, we hold onto a ref to the scalingMode that persists across renders. When VideoViewOptions.scalingMode
+  // differs from the persistent ref, we know to call updateScalingMode. We then subsequently update the ref to the new value for
+  // future renders.
+  //
+  // For scenario 2, we extract the VideoViewOptions.isMirrored to ensure any change to the isMirrored prop triggers the useEffect
+  // that recreates the stream view. We must also here add an extra check for when the scaling mode changes and we do not have
+  // access to the original StreamView (and hence cannot call updateScalingMode). When this happens we must also trigger the useEffect
+  // that recreates the stream view.
+
+  const [streamRendererResult, setStreamRendererResult] = React.useState<CreateVideoStreamViewResult>();
+  const scalingModeRef = useRef(scalingMode);
+  const hasScalingModeChanged = scalingModeRef.current !== scalingMode;
+  const updatingScalingModeDirectly = hasScalingModeChanged && !!streamRendererResult;
+
+  if (isStreamAvailable && renderElementExists && scalingMode && updatingScalingModeDirectly) {
+    (async () => {
+      streamRendererResult && (await streamRendererResult.view.updateScalingMode(scalingMode));
+      scalingModeRef.current = scalingMode;
```

Comment:
Pushing on this a bit more. Thinking through various scenarios that can happen while `updateScalingMode` is blocked.

- We rerender without any changes to props at all (for this hook at least) -- we'll end up calling `updateScalingMode` again (because line 45 will result in `true`)
- stream is disposed (is this possible?): we might end up setting `scalingModeRef.current` to an incorrect value and then have the wrong `scalingMode` going forward. - I'm not a 100% sure of this, but would like to talk this through and document (in comments) why it is not possible.

Reply:
For number 1 - nice catch -- may need an isUpdatingRef check (gross), will look into this.

For number 2:
a. stream is disposed through onComponentDismount -- not concerned about this as component has dismounted
b. stream is disposed because isMirrored or one of the other useEffect dependencies on L85 changes -- another nice catch -- this needs the below useEffect to signal this code block to not update the scalingModeRef... which sounds messy.. I'll have a look into this

Reply:
I could set `scalingModeRef.current = scalingMode` first, but then on failure of updateScalingMode it would be incorrect